packages/jest-console/src/matchers.js

```diff
@@ -8,57 +8,72 @@ import { matcherHint, printExpected, printReceived } from 'jest-matcher-utils';
  */
 import supportedMatchers from './supported-matchers';
 
+const createErrorMessage = ( spyInfo ) => {
+	const { spy, pass, calls, matcherName, methodName, expected } = spyInfo;
+	const hint = pass ? `.not${ matcherName }` : matcherName;
+	const message = pass
+		? `Expected mock function not to be called but it was called with:\n${ calls.map(
+				printReceived
+		  ) }`
+		: `Expected mock function to be called${
+				expected ? ` with:\n${ printExpected( expected ) }\n` : '.'
+		  }\nbut it was called with:\n${ calls.map( printReceived ) }`;
+
+	return () =>
+		`${ matcherHint( hint, spy.getMockName() ) }` +
+		'\n\n' +
+		message +
+		'\n\n' +
+		`console.${ methodName }() should not be used unless explicitly expected\n` +
+		'See file://gutenberg/packages/jest-console/README.md for details.';
+};
+
+const createSpyInfo = ( spy, matcherName, methodName, expected ) => {
+	const calls = spy.mock.calls;
+
+	const pass = expected
+		? JSON.stringify( calls ).includes( JSON.stringify( expected ) )
+		: calls.length > 0;
+
+	const message = createErrorMessage( {
+		spy,
+		pass,
+		calls,
+		matcherName,
+		methodName,
+		expected,
+	} );
+
+	return {
+		calls,
+		pass,
+		message,
+	};
+};
+
 const createToHaveBeenCalledMatcher =
 	( matcherName, methodName ) => ( received ) => {
 		const spy = received[ methodName ];
-		const calls = spy.mock.calls;
-		const pass = calls.length > 0;
-		const message = pass
-			? () =>
-					matcherHint( `.not${ matcherName }`, spy.getMockName() ) +
-					'\n\n' +
-					'Expected mock function not to be called but it was called with:\n' +
-					calls.map( printReceived )
-			: () =>
-					matcherHint( matcherName, spy.getMockName() ) +
-					'\n\n' +
-					'Expected mock function to be called.';
+		const spyInfo = createSpyInfo( spy, matcherName, methodName );
 
 		spy.assertionsNumber += 1;
 
 		return {
```

Comment:
It looks like it could get simplified to `return spyInfo;` if the `spyInfo` would contain only the message and pass fields.